src/main/java/org/spongepowered/common/event/SpongeCommonEventFactory.java

```diff
@@ -281,6 +285,43 @@ public static CollideEntityEvent callCollideEntityEvent(net.minecraft.world.Worl
         return event;
     }
 
+    // for saner debugging
+    public static boolean handlePistonEvent(IMixinWorldServer world, WorldServer.ServerBlockEventList list, Object obj, BlockPos pos, Block blockIn, int eventId, int eventParam) {
+        boolean extending = (eventId == 0);
+        final IBlockState blockstate = ((net.minecraft.world.World)world).getBlockState(pos);
+        EnumFacing direction = (EnumFacing) blockstate.getValue(BlockDirectional.FACING);
+
+        List<BlockPos> blocks = new ArrayList<BlockPos>();
+        blocks.add(pos);
+
+        BlockPistonStructureHelper movedBlocks = new BlockPistonStructureHelper((WorldServer) (Object) world, pos, direction, extending);
+        movedBlocks.canMove(); // calculates blocks to be moved
+
+        List<BlockPos> blocksToAdd = new ArrayList<BlockPos>();
+        blocksToAdd.addAll(movedBlocks.getBlocksToDestroy());
+        blocksToAdd.addAll(movedBlocks.getBlocksToMove());
+        for (BlockPos add : blocksToAdd) { // filter out duplicates
```

Comment:
You're doing this unnecessarily complicated filter... you can simplify this down to a simple for loop of the movedBlocks.getBlocksToMove while creating new Location objects.